Fix Airflow serialization for namedtuple

Joffreybvn opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

This PR fixes: #36839
In some cases, the namedtuple was not recognized as tuple and failed to serialize.


<br>

__Side note__:
When we return a tuple from a task, the tuple reach the XComEncoder and get serialized as a dict - thanks to [this line](https://github.com/apache/airflow/blob/46470aba68e5ebeee24a03dc22d012a50ee287ad/airflow/utils/json.py#L101). However, when we return a list of tuples, the tuples get serialized as lists. The [above condition is false](https://github.com/apache/airflow/blob/46470aba68e5ebeee24a03dc22d012a50ee287ad/airflow/utils/json.py#L101), and we fallback to the default json encoding behavior. Thus, the way namedtuple / tuple are handled is inconsistent.

This causes some queries of the Databricks Hook and Operator to fail right now, depending on how the Hook / Operator are parameterized.

This PR fix the namedtuple for such queries, but it doesn't fix the inconsistency: sometimes namedtuple will be encoded as dict, sometimes as list. IMO a definitive fix involve another strategy than [`qualname()`](https://github.com/apache/airflow/blob/46470aba68e5ebeee24a03dc22d012a50ee287ad/airflow/utils/module_loading.py#L47). Because namedtuples have dynamic qualnames, wich will never be "builtins.tuple".

<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

joffreybienvenu-infrabel committed:
feat: Implement namedtuple serialization

namedtuple is serialized like 'builtins.tuple'

joffreybienvenu-infrabel committed:
fix: static-checks

bolkedebruin commented on airflow/serialization/serde.py:
```diff
@@ -147,6 +147,13 @@ def serialize(o: object, depth: int = 0) -> U | None:
+    # serialize namedtuple like tuples
+    if _is_namedtuple(o):
+        qn = "builtins.tuple"
+        data, _, version, is_serialized = _serializers[qn].serialize(o)
```
This is redundant. Setting qn to "builtins.tuple" and move this check higher is sufficient.

Joffreybvn replied:
Indeed, refactored. Thanks !

bolkedebruin commented on airflow/serialization/serde.py:
```diff
@@ -337,6 +344,11 @@ def _is_pydantic(cls: Any) -> bool:
 
+def _is_namedtuple(cls: Any) -> bool:
+    """Return True if the class is a namedtuple."""
+    return isinstance(cls, tuple) and hasattr(cls, "_asdict") and hasattr(cls, "_fields")
```
isinstance is slow. Is it possible to just check on attributes?

Joffreybvn replied:
Yes, done

bolkedebruin reviewed (changes requested):
I'm picky in this area due for speed and tightness. This gets called *a lot*.  Make sure to check for attributes as much as you can (hasattr > isinstance)

joffreybienvenu-infrabel committed:
fix: check namedtuple based on attributes

In the future, if another object very similar to tuple exists, we can also check for `_make` and `_replace` attributes.

joffreybienvenu-infrabel committed:
fix: refactor duplicates serialization block

The classname of the namedtuple has to be `"builtins.tuple"`, instead of the qualname (the qualname will refer to the name given to the namedtuple when created).

I assume that `_serializers[qn].serialize(o)` won't always return a `classname` == qualname. Thus, creating a new `classname` variable set to "builtins.tuple", to ensure the namedtuple is considered as tuple.

joffreybienvenu-infrabel committed:
fix: put custom serializer block after builtin serializers

bolkedebruin commented on airflow/serialization/serde.py:
```diff
@@ -134,19 +134,25 @@ def serialize(o: object, depth: int = 0) -> U | None:
 
     cls = type(o)
     qn = qualname(o)
+    classname = None
```
This is not required. Please remove it.

bolkedebruin commented on airflow/serialization/serde.py:
```diff
@@ -134,19 +134,25 @@ def serialize(o: object, depth: int = 0) -> U | None:
+    # serialize namedtuple like tuples
+    if _is_namedtuple(o):
+        qn = "builtins.tuple"
+        classname = qn
```
It does not need to set the classname. Just the qualified name.

bolkedebruin commented on airflow/serialization/serde.py:
```diff
@@ -134,19 +134,25 @@ def serialize(o: object, depth: int = 0) -> U | None:
+        qn = "builtins.tuple"
+        classname = qn
+
+    # if there is a builtin serializer available use that
```
Keep this below. Do not move it.

bolkedebruin commented on airflow/serialization/serde.py:
```diff
@@ -337,6 +343,15 @@ def _is_pydantic(cls: Any) -> bool:
+    Checking is done by attributes as it is significantly faster than
+    using isinstance.
+    """
+    return hasattr(cls, "_asdict") and hasattr(cls, "_fields") and hasattr(cls, "_field_defaults")
```
Awesome.

bolkedebruin reviewed (changes requested):
Please keep the changes to a minimum. The only two lines that are required are

```python
if _is_namedtuple(o):
  qn = "builtins.tuple"
```

bolkedebruin committed:
Update serde.py

joffreybienvenu-infrabel committed:
fix: remove `classname`

joffreybienvenu-infrabel committed:
Revert "fix: remove `classname`"

This reverts commit e7be46531876f4085c6784dd2a23af2cecdca3a3.

joffreybienvenu-infrabel committed:
Revert "Update serde.py"

This reverts commit 9015e3ad78681e171083c524b5ddafde049740f9.

joffreybienvenu-infrabel committed:
fix: Add comment on classname overriding

Note: we do need to specify the classname, which should be "builtins.tuple":

The serialize method (in airflow/serialization/serializers/builtin.py) does qualname() on the namedtuple, which returns "airflow.providers.databricks.hooks.Row" (the namedtuple dynamically created in the hook). If this is used as classname, it will fail to deserialize: there won't be any deserializer for it.

If we hardcode it to "builtins.tuple", it gets deserialized correctly.
